### ISSUE TITLE
Package request: TensorRT

### ISSUE THREAD

**Issue by hmaarrfk** [conda-forge/staged-recipes#25661]
### Package name

tensorrt

### Package version

Newest

### Package website

https://github.com/NVIDIA/TensorRT
https://pypi.org/project/tensorrt/

### Package availability

https://pypi.org/project/tensorrt/

### Additional comments

It seems that in 2022 this was taken off the cuda build list:
https://github.com/conda-forge/staged-recipes/issues/21382

any particular reason for that? Was the github page for TensorRT missing at the time??

cc: @jakirkham 

### Package is not available

- [X] The package is not available on conda-forge.

### No previous issues or open PRs

- [X] No previous issue exists and no PR has been opened.

**Comment by jakirkham** [comment]
Thanks for raising Mark! 🙏 

<hr>

First a small clarification on this point:

> It seems that in 2022 this was taken off the cuda build list:
> https://github.com/conda-forge/staged-recipes/issues/21382

It's still on the list. It is unchecked with the note that it is missing a redist

<img width="858" alt="Screenshot 2024-03-14 at 12 32 24 PM" src="https://github.com/conda-forge/staged-recipes/assets/3019665/68745b4e-e3a0-4e68-9a8b-c6d81ca6b2a5">

<hr>

To provide more context on that, the CUDA packages are built in an internal pipeline that creates a binary redist at the end. Currently we lack that process for TensorRT. So that will need to be built out first. This may take some time

As to the GitHub repo, this is providing only the open source portion, which includes things like examples, samples, some open source code, etc.. However it is still missing the closed source portions, which are not distributed there. Without those, it probably won't be sufficient

AFAICT the PyPI artifact is just an `sdist` from the GitHub repo. So similarly lacks the closed source portions. Presumably a wheel would be needed with the relevant binary components

<hr>

In any event have followed up internally and made sure we have an issue tracking this need. Will update the old CUDA listing issue to reference this issue

If you are able to share more about your use cases in this issue, that would be helpful. Are there particular packages that would benefit from building with TensorRT? What use cases do you or others have in mind where TensorRT enabled builds would help?

**Comment by hmaarrfk** [comment]
Thx I saw the note and was confused about  how the stance might have changed, but I wanted to revive the issue.

It would be great to have onnxruntime enabled TensorRT. I believe pytorch and tensorflow now have the ability to leverage TensorRT as well.

Onnx has proved to be amazing for ML deployment, so we can plug in TensorRT when the hardware is available.

We've shown that TensorRT for our usecases would create maybe 30% boost in our inference models, but creating a dedicated model for TensorRT might be too much churn for our team. I would rather just have an onnx model and let the client optimize as needed.

**Comment by jakirkham** [comment]
Thanks for doing that! 🙏

This also helps signal what is important. NVIDIA has a lot of software. So having input from users about what is needed is helpful

Appreciate the insight. So is it this issue ( https://github.com/conda-forge/onnxruntime-feedstock/issues/109 )? Or is there a different issue that is relevant?

**Comment by hmaarrfk** [comment]
> Appreciate the insight. So is it this issue ( https://github.com/conda-forge/onnxruntime-feedstock/issues/109 )? Or is there a different issue that is relevant?

Yes. it is relevant. Truthfully, I opened both issues since it seems that tensorrt is both a C++ library (potentially included by the operating system???? package) and a python library.

> NVIDIA has a lot of software. So having input from users about what is needed is helpful

I totally understand that. Your efforts here are definitely not ignored by me and the rest of my team!

From a system standpoint and "edge deployment", the more performance we can edge out of our GPUs, the more likely we are to use them. The engineers on our team have shown that in many cases running the models on the CPUs gives "good enough" performance. Having TensorRT give us a 30% boost in our performance means that we can think of running the models for "real time" use cases, but you know, only when we buy the expensive GPUs ;).

**Comment by traversaro** [comment]
fyi @carloscp3009 this is the tracking issue I was referring to.

**Comment by ehfd** [comment]
TensorRT-LLM might be worth a look as well along the way.

**Comment by jakirkham** [comment]
Thanks Seungmin! 🙏

Please feel free to add that as a new issue and link it to this one

It is good to hear these use cases. Just want to make sure we also are tracking these downstream package requests too

**Comment by ehfd** [comment]
Linked with https://github.com/conda-forge/staged-recipes/issues/28973.

**Comment by ehfd** [comment]
I guess https://docs.nvidia.com/deeplearning/tensorrt/latest/installing-tensorrt/installing.html#tar-file-installation is the way to repackage this distribution?

**Comment by diegoferigo-rai** [comment]
Just checking in, is there any update on this package request?

**Comment by traversaro** [comment]
xref: https://github.com/conda-forge/staged-recipes/pull/34373 .

**Comment by jakirkham** [comment]
Thanks all for your interest here! 🙏

Brought this up in our weekly standup.

Looks like we now have a `redist` (a necessary precondition for a Conda package).

Will be looking with our team into whether that works for our Conda packaging needs.

**Comment by jakirkham** [comment]
Completed with PR: https://github.com/conda-forge/staged-recipes/pull/34373

Can now be found in this feedstock: https://github.com/conda-forge/tensorrt-feedstock
